### PR TITLE
HF radar radial velocity converter

### PR DESCRIPTION
**Pull request by lliu1220**
## Description

This PR was to enable HF marine radar radial velocity from archiving to be converted into ioda format for marine JEDI to be able to assimilate radar data.

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes 339: https://github.com/JCSDA-internal/ioda-converters/issues/339

## Acceptance Criteria (Definition of Done)

This PR shall conform to ioda-converters code norm;
This PR shall be able to convert testing HF radar data 20200720.nc into ioda format.

## Dependencies

This PR does not have dependency

## Impact

Please list the other repositories (if any) that this PR will require changes in (example below)

- soca-science
- Next step is to add configuration into soca-science to enable workflow assimilating HF radar

## Test Data

Please list the test data (if any) that needs to be updated on AWS to the best of your knowledge (example below).

- data need to be obtained from NOAA archiving,

### CODE REVIEW

**Comment by srherbener** [comment]
The tests are failing due to coding norms errors:
```
test 5
--
933 | Start 5: iodaconv_marine_coding_norms
934 |  
935 | 5: Test command: /build_container/bin/iodaconv_lint.sh "/jcsda/ioda-bundle/iodaconv/src/marine" "/jcsda/ioda-bundle/iodaconv"
936 | 5: Environment variables:
937 | 5:  OMP_NUM_THREADS=1
938 | 5: Test timeout computed to be: 10000000
939 | 5: + SRC_DIR=/jcsda/ioda-bundle/iodaconv/src/marine
940 | 5: + PYCODESTYLE_CFG_PATH=/jcsda/ioda-bundle/iodaconv
941 | 5: + cd /jcsda/ioda-bundle/iodaconv/src/marine
942 | 5: + pycodestyle -v --config=/jcsda/ioda-bundle/iodaconv/.pycodestyle --filename=*.py,*.py.in .
943 | 5: cli configuration: /jcsda/ioda-bundle/iodaconv/.pycodestyle
944 | 5: directory .
945 | 5: checking ./argoClim2ioda.py
946 | 5: checking ./cryosat_ice2ioda.py
947 | 5: checking ./emc_ice2ioda.py
948 | 5: checking ./gds2_sst2ioda.py
949 | 5: checking ./glider2ioda.py
950 | 5: checking ./gmao_obs2ioda.py
951 | 5: checking ./godae_profile2ioda.py
952 | 5: checking ./godae_ship2ioda.py
953 | 5: checking ./godae_trak2ioda.py
954 | 5: checking ./hgodas_adt2ioda.py
955 | 5: checking ./hgodas_insitu2ioda.py
956 | 5: checking ./hgodas_sst2ioda.py
957 | 5: checking ./radarmarine_obs2ioda.py
958 | 5: ./radarmarine_obs2ioda.py:3:54: W291 trailing whitespace
959 | 5: ./radarmarine_obs2ioda.py:37:1: E302 expected 2 blank lines, found 1
960 | 5: ./radarmarine_obs2ioda.py:55:38: W291 trailing whitespace
961 | 5: ./radarmarine_obs2ioda.py:58:13: E231 missing whitespace after ','
962 | 5: ./radarmarine_obs2ioda.py:58:37: E231 missing whitespace after ','
963 | 5: ./radarmarine_obs2ioda.py:63:37: E231 missing whitespace after ','
964 | 5: ./radarmarine_obs2ioda.py:63:47: E231 missing whitespace after ','
965 | 5: ./radarmarine_obs2ioda.py:63:50: W291 trailing whitespace
966 | 5: ./radarmarine_obs2ioda.py:66:1: E115 expected an indented block (comment)
967 | 5: ./radarmarine_obs2ioda.py:67:11: E111 indentation is not a multiple of four
968 | 5: ./radarmarine_obs2ioda.py:67:22: E231 missing whitespace after ','
969 | 5: ./radarmarine_obs2ioda.py:75:15: E111 indentation is not a multiple of four
970 | 5: ./radarmarine_obs2ioda.py:76:15: E111 indentation is not a multiple of four
971 | 5: ./radarmarine_obs2ioda.py:77:15: E111 indentation is not a multiple of four
972 | 5: ./radarmarine_obs2ioda.py:79:15: E111 indentation is not a multiple of four
973 | 5: ./radarmarine_obs2ioda.py:80:15: E111 indentation is not a multiple of four
974 | 5: ./radarmarine_obs2ioda.py:81:15: E111 indentation is not a multiple of four
975 | 5: ./radarmarine_obs2ioda.py:83:1: E302 expected 2 blank lines, found 1
976 | 5: checking ./rads_adt2ioda.py
977 | 5: checking ./smap_sss2ioda.py
978 | 5: checking ./viirs_modis_oc2ioda.py
979 | 5: 7       E111 indentation is not a multiple of four
980 | 5: 1       E115 expected an indented block (comment)
981 | 5: 5       E231 missing whitespace after ','
982 | 5: 2       E302 expected 2 blank lines, found 1
983 | 5: 3       W291 trailing whitespace
984 | 1/1 Test #5: iodaconv_marine_coding_norms .....***Failed    0.33 sec
985 |  
986 | 0% tests passed, 1 tests failed out of 1
987 |  
988 | Label Time Summary:
989 | iodaconv    =   0.33 sec*proc (1 test)
990 | script      =   0.33 sec*proc (1 test)
991 |  
992 | Total Test time (real) =   0.33 sec
993 |  
994 | The following tests FAILED:
995 | 5 - iodaconv_marine_coding_norms (Failed)
996 | Errors while running CTest
```

**Comment by lliu1220** [comment]
@srherbener 
https://github.com/JCSDA-internal/ioda-converters/issues/480

**Review by guillaumevernieres — CHANGES_REQUESTED** [review]
Follow the naming convention that was adopted:
Rename `radarmarine_obs2ioda.py` to `<provider>_<obs_type>2ioda.py`
and as @srherbener pointed out, these are your code norm issues:
```console
958 | 5: ./radarmarine_obs2ioda.py:3:54: W291 trailing whitespace
959 | 5: ./radarmarine_obs2ioda.py:37:1: E302 expected 2 blank lines, found 1
960 | 5: ./radarmarine_obs2ioda.py:55:38: W291 trailing whitespace
961 | 5: ./radarmarine_obs2ioda.py:58:13: E231 missing whitespace after ','
962 | 5: ./radarmarine_obs2ioda.py:58:37: E231 missing whitespace after ','
963 | 5: ./radarmarine_obs2ioda.py:63:37: E231 missing whitespace after ','
964 | 5: ./radarmarine_obs2ioda.py:63:47: E231 missing whitespace after ','
965 | 5: ./radarmarine_obs2ioda.py:63:50: W291 trailing whitespace
966 | 5: ./radarmarine_obs2ioda.py:66:1: E115 expected an indented block (comment)
967 | 5: ./radarmarine_obs2ioda.py:67:11: E111 indentation is not a multiple of four
968 | 5: ./radarmarine_obs2ioda.py:67:22: E231 missing whitespace after ','
969 | 5: ./radarmarine_obs2ioda.py:75:15: E111 indentation is not a multiple of four
970 | 5: ./radarmarine_obs2ioda.py:76:15: E111 indentation is not a multiple of four
971 | 5: ./radarmarine_obs2ioda.py:77:15: E111 indentation is not a multiple of four
972 | 5: ./radarmarine_obs2ioda.py:79:15: E111 indentation is not a multiple of four
973 | 5: ./radarmarine_obs2ioda.py:80:15: E111 indentation is not a multiple of four
974 | 5: ./radarmarine_obs2ioda.py:81:15: E111 indentation is not a multiple of four
975 | 5: ./radarmarine_obs2ioda.py:83:1: E302 expected 2 blank lines, found 1
```
You also need to add a unit test somewhere around [here](https://github.com/JCSDA-internal/ioda-converters/blob/e8c0b20d18ad7b95e55aa676e189e60f5d9bf84a/test/CMakeLists.txt#L219)

**Comment by lliu1220** [comment]
@guillaumevernieres 
There is issue compile ioda-converters,
I'm fixing code norm, but it won't compile or merge
see my issue https://github.com/JCSDA-internal/ioda-converters/issues/480

**Comment by guillaumevernieres** [comment]
> @guillaumevernieres
> There is issue compile ioda-converters,
> I'm fixing code norm, but it won't compile or merge
> see my issue #480

Just fix your coding norm and add a unit test for now.

**Comment by lliu1220** [comment]
ok. @guillaumevernieres

File: test/CMakeLists.txt
Comment (liuxiao37k) [inline test/CMakeLists.txt:60]: I believe you would also need to add ndbc_hfradar.nc to the /testoutput folder, like what you did for /testinput, so that it can be compared with the file generated in /testrun.

**Comment by liuxiao37k** [comment]
The new ctest failed on Hera. I believe you would need to add `ndbc_hfradar.nc` to both /testinput (which you did) and /testoutput (which you didn't) folders.
```
20/49 Test #20: test_iodaconv_hgodas_insitu ..........   Passed    0.39 sec
      Start 21: test_iodaconv_hgodas_sst
21/49 Test #21: test_iodaconv_hgodas_sst .............   Passed    0.40 sec
      Start 22: test_iodaconv_glider
22/49 Test #22: test_iodaconv_glider .................   Passed    4.31 sec
      Start 23: test_iodaconv_hfradar
23/49 Test #23: test_iodaconv_hfradar ................***Failed    8.32 sec
      Start 24: test_iodaconv_argoclim
24/49 Test #24: test_iodaconv_argoclim ...............   Passed    0.74 sec
      Start 25: test_iodaconv_cryosat2
25/49 Test #25: test_iodaconv_cryosat2 ...............   Passed    0.63 sec
      Start 26: test_iodaconv_metar
26/49 Test #26: test_iodaconv_metar ..................   Passed    0.45 sec
      Start 27: test_iodaconv_viirs_jpss1_oc_l2
27/49 Test #27: test_iodaconv_viirs_jpss1_oc_l2 ......   Passed    2.51 sec
```
more details:
```
23: Test command: .../iodaconv_comp.sh "netcdf" ".../ndbc_hfradar2ioda.py
                          -i testinput/ndbc_hfradar.nc
                          -o testrun/ndbc_hfradar.nc
                          -d 2016080412" "ndbc_hfradar.nc" "0.0"
23: Environment variables: 
23:  OMP_NUM_THREADS=1
23: Test timeout computed to be: 10000000
23: 2021-03-23 19:58:01.305276 +0000 ERROR ../../src/nccmp_state.c:156 No such file or directory
1/1 Test #23: test_iodaconv_hfradar ............***Failed    8.25 sec
```

**Review by liuxiao37k — CHANGES_REQUESTED** [review]

**Comment by liuxiao37k** [comment]
> > @guillaumevernieres
> > There is issue compile ioda-converters,
> > I'm fixing code norm, but it won't compile or merge
> > see my issue #480
> 
> Just fix your coding norm and add a unit test for now.

Just a tip - I also experienced some issue building ioda-converter on Hera (as of this afternoon), but I just skipped the bufr ctests and ran the rest of the ctests.

**Comment by lliu1220** [comment]
@liuxiao37k 
thanks for tips, Xiao!
right now, ioda-converters or ioda-bundles develop branch has compile issue, the code norms for HF radar have been corrected, but I will try what you suggested to comment off the bufr_ctest, it does seem to be the issue since my serial compile without parallel indicates the compile error is related to bufr and it is on Orion, so good to know you ran into same issue on Hera.

**Comment by jkbk2004** [comment]
@lliu1220 @liuxiao37k I noticed that jedi-stack issue on hera was reported to core team.

**Comment by liuxiao37k** [comment]
> @lliu1220 @liuxiao37k I noticed that jedi-stack issue on hera was reported to core team.

Yes. The workaround to the jedi-stack issue on Hera is to build soca and soca-science with the oops branch `feature/work_around_hera_once_more` until it is merged to `develop`.

**Comment by lliu1220** [comment]
fixed compile and passed radar ctest on Orion:

    Start 21: test_iodaconv_hfradar

21: Test command: /work/noaa/marine/lliu/ioda/ioda_0322/ioda-converters/build/bin/iodaconv_comp.sh "netcdf" "/work/noaa/marine/lliu/ioda/ioda_0322/ioda-converters/build/bin/ndbc_hfradar2ioda.py
                          -i testinput/ndbc_hfradar_in.nc
                          -o testrun/ndbc_hfradar_out.nc
                          -d 2020072012" "ndbc_hfradar_out.nc" "0.0"
21: Environment variables:
21:  OMP_NUM_THREADS=1
21: Test timeout computed to be: 1500
1/1 Test #21: test_iodaconv_hfradar ............   Passed    5.85 sec

The following tests passed:
        test_iodaconv_hfradar

100% tests passed, 0 tests failed out of 1

Label Time Summary:
iodaconv    =   5.85 sec*proc (1 test)
script      =   5.85 sec*proc (1 test)

Total Test time (real) =   5.86 sec

**Comment by lliu1220** [comment]
@srherbener @guillaumevernieres @liuxiao37k 
waiting on ci-test to pass. thank you for all of you who commented and reviewed this PR

**Comment by srherbener** [comment]
Unfortunately, the CI testing is broken now. The core team is working to get this fixed. I just tested this on my iMac and all the tests pass including the new radar test. So functionally we are in good shape with this PR. Thank you @lliu1220 for your patience as we work through these issues.

**Review by srherbener — CHANGES_REQUESTED** [review]
The test file is way too big for the purpose of unit testing. It contains 85,878 locations (nlocs) and we are typically wanting under 100 locations for unit testing. The idea behind unit testing is to exercise as much of the code as possible with a very short, fast running test.

Scientific validation can be done with a full set of obs outside of unit testing. We are developing a tiered testing strategy where tier 1 is for unit testing, and tier 2 and above are for scientific validation. At this point, the tier 2 and above testing needs to be done manually outside of the ctest, but eventually these tests will be integrated into the ctest scheme.

The test file you are using now is okay for tier 2 and above, but it would be good to trim down the size of the file for the unit testing in this PR. You may be able to use NCO utilities to do this. check out the -d option on ncks from the NCO package.

Thank you for your help with this!

**Comment by lliu1220** [comment]
@srherbener 

Let me know if this is ok:
the current input data is one sample data piece from NDBC, it contains mostly empty place holder, in the python script I just committed: added a line to filter out the vast majority of empty data: if vals_u[i] != '--':
Thus it left us with only about 8000 data as output: ${ioda-converters_dir}/test/testoutput/ndbc_hfradar_out.nc
Thank you.

**Comment by srherbener** [comment]
@lliu1220 thank you for trimming down the test file. Unfortunately, the test is still taking about 7 sec on my iMac. Ideally, we would like the test time to be under 2 sec (but I realize that there are other tests that take more than 2 sec). 8000 locations is still a pretty large number for unit testing.

Are you saying that you are at the limit of trimming the data because the input is a single sample? It would be fine to use a partial sample for unit testing because the goal is to exercise code and not to have a complete valid data set (in a scientific sense). The testing of a valid data set can be done in the tier2 and above testing.

As an experiment, I tried doing this using the file checked into your branch:
```
module load nco
ncks -d lat,100,150,5 -d lon,100,150,5 iodaconv/test/testinput/ndbc_hfradar_in.nc test.nc
bin/ndbc_hfradar2ioda.py -i test.nc4 -o test.out.nc4 -d 2020072012
```
This produced an output file with 61 locations, and the converter ran in about 0.5 sec. This is probably enough to get decent code coverage and it runs fast. Would this work?

**Comment by lliu1220** [comment]
@srherbener 
thank you. did what you suggested.

**Review by srherbener — APPROVED** [review]
@lliu1220 thank you very much for adding in a new test and trimming down the test data. Looks great now!

**Review by liuxiao37k — APPROVED** [review]

**Comment by srherbener** [comment]
@guillaumeverniere,have your comments been addressed? If so, please approve to clear out the "change requests". Thanks!

**Review by YanHao-NOAA — APPROVED** [review]

**Comment by srherbener** [comment]
@guillaumevernieres requested changes were:

- change name of the script to the `<provider>_<obstype>2ioda.py`
- add a unit test

It appears that both of these have been addressed, so I am going to merge this in now.

**Comment by lliu1220** [comment]
@srherbener have a question on how ioda2 new change will affect amending the following PR, current reader read out one-dimensional array:
Reopened this branch cause there are a few things need to be amended.
For instance the current converter convert velocity into 
image
but they should be after amendment
image
the problem is line 64 has flattened a 3d array in a wrong way that time dimension was kneaded into geo dimensions
vals_u = vals_u.flatten()
it is ok for when time==1, but not ok when time dimension size > 1
Thank you.

**Comment by srherbener** [comment]
The upgrade of the ioda-converters to produce ioda-v2 files has been deferred until after the next JEDI release (scheduled for Jun 11). We are planning a code sprint for upgrading the converters likely to take place in mid July this summer.

@CoryMartin-NOAA is working on upgrading the library you are using (BuildNetcdf call) to write out in ioda-v2 format which will be introduced during the code sprint.

For now, the solution is to have the converters continue to write ioda-v1 files, and use the ioda upgrader application to convert their output to ioda-v2. The upgrader is provided when building a bundle (ioda-bundle, ufo-bundle, etc) and appears in `your_build_area/bin/ioda-upgrade.x`. The usage is:

```
ioda-upgrade.x ioda-v1-file ioda-v2-file
```

What I'm thinking is that you can do your amendments now in the context of writing out 1D variables into ioda-v1 format (and then use the upgrader to convert that to ioda-v2 format for testing). Then during the code sprint you can figure out how to modify your converter to accommodate multi-dimensioned variables. If this doesn't work, then let me know and I can help out.

Thanks!

**Pull request merged**
